feat: Add ADR 284 - Private voice chat (#299)
decentraland/adr: content/ADR-284-private-voice-chat.md

---
layout: adr
adr: 284
date: 2025-06-30
title: Private voice chat
authors:
  - LautaroPetaccio
  - pentreathm
status: Review
type: Standards Track
spdx-license: CC0-1.0
---

## Abstract

This specification defines the private voice chat protocol, designed to provide voice communication between two users in the Explorer, simulating a two person call.

## Context

After having had the voice chat for nearby users in previous versions of Decentraland, the need of having private voice conversations between users was identified as an important feature.

In this new feature, a user (now on called the **caller**) is able to ring another user (from now on the **callee**), which initiates the call flow.
The callee has the capability of rejecting this call or accepting it and the caller can cancel the process at any time.

> If the callee chooses to reject it or the caller cancels it, the ringing process will end, notifying the caller about the decision the callee made.

Reviewer: wdyt?
```suggestion
If the callee rejects the call or the caller cancels it, the ringing process ends, notifying the user of the action taken by the other party.
```

Author: Looks better! Changed it!

> Case contrary, by accepting the call, the callee and the caller will both receive tokens to connect to the real time service that will be providing the service for the call ([LiveKit](https://livekit.io/) at the time of this ADR).

Reviewer 2: ```suggestion
Case contrary, by accepting the call, the callee and the caller will both receive tokens to connect to the real time service that will facilitate the call ([LiveKit](https://livekit.io/) at the time of this ADR).
```

> Both users can end the call at any time, which will result in the other user being disconnected from the call.

Reviewer: ```suggestion
Either user can end the call at any time, disconnecting the other and finalizing the call.
```

Author: Changed it!

The feature considers a some restrictions, with the aim to improve the user experience, that need to be enforced:

> - A user can't be in more than one private voice chat at a time.

Reviewer 2: ```suggestion
- A user cannot participate in more than one private voice chat simultaneously
```

- A user can't be ringing or be ringed by another user at the time.

> - A user can't start a private voice chat with another user if at least one of them has their privacy settings set to _only friends_ and their are not friends.

Reviewer 2: ```suggestion
- A user can't start a private voice chat with another user if at least one of them has their privacy settings set to _only friends_ and they are not friends.
```

Author: Changed it!

> - A user can't be ringing someone or be ringed be someone if they're not connected.

Reviewer 2: ```suggestion
- A user cannot initiate a call or receive a call from someone if they are not connected.
```

> - A user can't be initiate a call or receive a call from be someone if they're not connected.

Reviewer: or to follow standard:

`A user can't initiate or receive a call unless it is connected in-world.`

> - A user can't be initiate a call or receive a call from be someone if they're not connected.

Author: I chose the second one to keep the consistency between them. Changed it!

Following the user restrictions mentioned above, there are some restrictions for the voice chat that are required to be enforced as well:

- A ringing private voice chat can't be ringing for an indeterminate amount of time, they must be expired.
- A private voice chat where one of the users didn't connect for an expected amount of time must not exist

> - A private voice chat where one of the users abruptly disconnected for an expected amount of time must not exist

Reviewer 2: Not sure I get this sentence, what does abruptly mean in this context? what should not exists is a call with a single user, right?

Author: You're right, this wasn't really clear. I've changed it for:
```
A private voice chat where one of the users disconnected, due to connectivity issues, for an expected amount of time must not exist
```

Reviewer 2: lgtm

- A private voice chat where one of the users disconnected willingly must not exist.

> All of these restrictions are enforced either through the Social Services or the Comms Gatekeeper, through private voice chat updates or through LiveKit actions respectively.

Author: Changed it!

## Specification

> This section specifies the interaction between the Explorer and the Social Service for the purposes of ringing someone and receiving voice chat updates and LiveKit for

Author: I've changed it a bit, what do you think about it now?

> This section specifies the interaction between the Explorer and the Social Service for the purposes of ringing someone, receiving voice chat updates and using LiveKit for
> the real time voice communication interactions.

Reviewer: ```suggestion
This section specifies the interaction between the Explorer and the Social Service for the purposes of ringing someone, receiving voice chat updates, and using LiveKit for real-time voice interactions.
```

### Starting a private voice chat

Starting a private voice chat is done by performing the `StartPrivateVoiceChat` RPC call with the callee address to the Social Service.

> If the start process is successful, that is, it complies with the requirements mentioned in the _Context_ section, the ringing process is started.

Author: Changed it!

> It must be taken into consideration that this is an asynchronous flow, meaning that if the start process is successful, it will return immediately with a call id.

Author: Changed it!

> This call ID can be used to identify the call and end it (if the user wants to). The callee will be notified through the updates subscription mechanism mentioned in the _Subscribing private voice chats updates_ section.

Author: Changed it!

If successful, it will return without errors and will end the ringing process. The caller will be notified through the updates subscription mechanism mentioned in the _Subscribing private voice chats updates_ section.

Here's the interaction flow of the Explorer and the Social Service:

```mermaid
sequenceDiagram
  actor U2 as Callee
  participant SS as Social Server

  U2->>SS: Reject call
  alt The call doesn't exist
	  SS-->>U2: Not found error
	else The call exists
	  SS-->>U2: Ok
	end
```

The RPC call is specified in the protocol as:

```protobuf
message RejectPrivateVoiceChatPayload {
  string call_id = 1;
}

message RejectPrivateVoiceChatResponse {
  message Ok {
    string call_id = 1;
  }

  oneof response {
    Ok ok = 1;
    InternalServerError internal_server_error = 2;
    NotFoundError not_found = 4;
  }
}

rpc RejectPrivateVoiceChat(RejectPrivateVoiceChatPayload) returns (RejectPrivateVoiceChatResponse) {}
```

### Accepting a private voice chat

Accepting a private voice chat is done by performing the `AcceptPrivateVoiceChat` RPC call to the Social Service with the call id of the ringing voice chat.

> If successful, it will return return the connection url, which is crafted using the LiveKit connection URL and the token to connect to the LiveKit room.

Reviewer: ```suggestion
If successful, it will return the connection URL, which is composed of LiveKit's connection URL and the token to connect to the LiveKit room.
```

Author: Changed it!

> Case contrary, the voice chat must be finished in the client, as this means that the voice chat was either expired or cancelled.

Reviewer: ```suggestion
Case contrary, the voice chat must be finished in the client, as this means that the voice chat had either expired or been cancelled.
```

Author: Changed it!

> This process will result in the caller will be notified through the updates subscription mechanism mentioned in the _Subscribing private voice chats updates_ section with the connection URL as well.

Reviewer: ```suggestion
This process will result in the caller being notified through subscription's updates mechanism mentioned in the (_Subscribing private voice chats updates_)[#Subscribing-private-voice-chats-updates] section with the connection URL as well.
```

> Upon receiving the connection URL, the client must connect to the LiveKit room using any implementation of the LiveKit's client. To communicate via voice, users must publish audio tracks in the room and process the audio tracks for which they're already subscribed. Check the [LiveKit's documentation](https://docs.livekit.io/home/) on how to do this.

Reviewer: ```suggestion
Upon receiving the connection URL, the client must connect to the LiveKit room using any LiveKit's client implementation. To communicate via voice, users must publish audio tracks in the room and process the audio tracks for which they're already subscribed. Check the [LiveKit's documentation](https://docs.livekit.io/home/) on how to do this.
```

```mermaid
sequenceDiagram
  actor U2 as Callee
  participant SS as Social Server
  participant LK as LiveKit

  U2->>SS: Accept private voice chat using the id
  alt The private voice chat doesn't exist
		SS-->>U2: Not found error
	else The private voice chat exists
		SS-->>U2: Connection URL
	end
  U2->>LK: Connect to the room
```

The RPC call is specified in the protocol as:

```protobuf
message AcceptPrivateVoiceChatPayload {
  string call_id = 1;
}

message AcceptPrivateVoiceChatResponse {
  message Ok {
    string call_id = 1;
    PrivateVoiceChatCredentials credentials = 2;
  }

  oneof response {
    Ok ok = 1;
    InternalServerError internal_server_error = 2;
    NotFoundError not_found = 4;
    ForbiddenError forbidden_error = 5;
  }
}

rpc AcceptPrivateVoiceChat(AcceptPrivateVoiceChatPayload) returns (AcceptPrivateVoiceChatResponse) {}
```

### Ending a private voice chat

Ending a private voice chat can be done through the Social Service in any of their states (ringing or on going) or through LiveKit if the user is already on an on going private voice chat.

#### Ending a private voice chat through the Social Service

> Ending a private voice chat is done by performing the `EndPrivateVoiceChat` RPC call with the call id of the voice chat.

Reviewer: ```suggestion
Ending a private voice chat is done by performing the `EndPrivateVoiceChat` RPC call to the Social Service with the call id of the voice chat.
```

Author: Changed it!

A voice chat can be ended at any state by any of the participants, but the callee should use the reject operation if they want to reject a private voice chat.

> This process will result in the caller will be notified through the updates subscription mechanism mentioned in the _Subscribing private voice chats updates_ section with the call id that is being terminated.

Reviewer: ```suggestion
This process will result in the caller being notified through the subscription's updates mechanism mentioned in the [_Subscribing private voice chats updates_](#Subscribing-private-voice-chats-updates) section with the call id that is being terminated.
```

Author: Changed it!

For each of these states the client must react to them in the following way:

- In the first case, the ringing process must be ended.
- In the second case, the voice chat must be ended by issuing a voluntarily disconnection from LiveKit.

Here's the interaction flow of the Explorer and the Social Service:

```mermaid
sequenceDiagram
	actor U1 as Caller or calle
	participant SS as Social Server
	participant LK as LiveKit

  U1->>SS: End private voice chat with id
  break Private voice chat exists
		SS-->>U1: Ok
	end
	LK->>U1: Room destroyed
```

```protobuf
message EndPrivateVoiceChatPayload {
  string call_id = 1;
}

message EndPrivateVoiceChatResponse {
  message Ok {
    string call_id = 1;
  }

  oneof response {
    Ok ok = 1;
    InternalServerError internal_server_error = 2;
    NotFoundError not_found = 3;
  }
}

rpc EndPrivateVoiceChat(EndPrivateVoiceChatPayload) returns (EndPrivateVoiceChatResponse) {}
```

#### Ending a private voice chat through LiveKit

An on going LiveKit private voice chat can be ended by issuing a voluntarily disconnection from the room.
This action is monitored by our services and will result in the LiveKit room being destroyed.

> The client must be listening to the LiveKit room events to detect the disconnection reaction.

Reviewer: ```suggestion
The client must be listening to the LiveKit room events to react upon disconnection.
```

In this case, the reason for the disconnection will be set as the destruction of the room, which the client must treat as the end of the private voice chat.

### Getting incoming private voice chats

> At least for now, we can't know if a user is connecter or not to the Social Service, therefore we can't strictly enforce the requirement to prevent users from starting private voice chats with disconnected users.

Reviewer: ```suggestion
At least for now, we can't know if a user is connected or not to the Social Service; therefore, we can't strictly enforce the requirement to prevent users from starting private voice chats with disconnected users.
```

Author: Changed it!

	alt Start private voice chat published
		SS-->>U1: Notify voice chat started with id and caller address
	else Private voice chat accepted published
		SS-->>U2: Notify voice chat accepted with credentials
	else Private voice chat rejected published
		SS-->>U2: Notify voice chat rejected with id
	else Private voice chat ended published
		alt Voice chat ended due to disconnection
			SS-->>U1: Notify voice chat ended with id
			SS-->>U2: Notify voice chat ended with id
		else Voice chat ended by callee
			SS-->>U2: Notify voice chat ended with id
		else Voice chat ended by caller
			SS-->>U1: Notify voice chat ended with id
		end
	else Private voice chat expired published
		SS-->>U1: Notify voice chat expired with id
		SS-->>U2: Notify voice chat expired with id
	end
```

Each update implies performing a different action, here is an explanation on how to act upon receiving them:

- **Voice chat started**: A caller started a conversation with the user. Start the ringing process in the client by using the address of the caller retrieved from the update and store the call id in case the user decides to reject or accept the private voice chat.

> - **Private voice chat accepted**: The callee accepted the private voice chat. Check that the call id coming with the update belongs to the private voice chat the caller created and then, connect to the LiveKit room using the connection url which comes with the update as well.

Reviewer: ```suggestion
- **Private voice chat accepted**: The callee accepted the private voice chat. Check that the call id coming from the update belongs to the private voice chat the caller has created, and then connect to the LiveKit room using the connection url which comes with the update as well.
```

Author: Changed it!

- **Private voice chat ended**: The caller or the callee ended the private voice chat. If the user is connected to LiveKit, issue a disconnection and end the voice chat. If the user has not connected yet, end the ringing process. This update can come from the intentional process of ending a call or from any of the users disconnecting from the Social Service.
- **Private voice chat expired**: The callee didn't answer the voice chat request. End the ringing process in the client.

The RPC call is specified in the protocol as:

```protobuf
message PrivateVoiceChatUpdate {
  string call_id = 1;
  PrivateVoiceChatStatus status = 2;
  optional User caller = 3;
  optional User callee = 4;
  optional PrivateVoiceChatCredentials credentials = 5;
}

rpc SubscribeToPrivateVoiceChatUpdates(google.protobuf.Empty) returns (stream PrivateVoiceChatUpdate) {}
```

### Recovering from a disconnection

Any user connected to a LiveKit room can experience an abrupt disconnection due to connectivity issues.
Our services will be monitoring the connection and disconnection of users from rooms.

> If a user were to be abruptly disconnected from the room as wasn't able to re-connect for a period of time, our service will issue the room to destroyed.

Reviewer: ```suggestion
If a user has been abruptly disconnected from the room and wasn't able to reconnect for a period of time, our service will issue the room as destroyed.
```
